### PR TITLE
feat(2690): make llm models configurable

### PR DESCRIPTION
**Pull request by beelchester**
fixes: #2690
/claim #2690

## Example
- Using Gemini 
Set `TAILCALL_LLM_API_KEY` to your Gemini API key.
```json
  "llm": {
    "model": "gemini-1.5-flash-latest",
    "secret": "{{.env.TAILCALL_LLM_API_KEY}}"
  }
```
- Using Ollama
Don't need secret.
```json
  "llm": {
    "model": "gemma2"
  }
```

## Models
```
--- Models for OpenAI
["gpt-4o", "gpt-4o-mini", "gpt-4-turbo", "gpt-4", "gpt-3.5-turbo"]

--- Models for Gemini
["gemini-1.5-pro", "gemini-1.5-flash", "gemini-1.0-pro", "gemini-1.5-flash-latest"]

--- Models for Anthropic
["claude-3-5-sonnet-20240620", "claude-3-opus-20240229", "claude-3-sonnet-20240229", "claude-3-haiku-20240307"]

--- Models for Groq
["llama-3.1-405b-reasoning", "llama-3.1-70b-versatile", "llama-3.1-8b-instant", "mixtral-8x7b-32768", "gemma-7b-it", "gemma2-9b-it", "llama3-groq-70b-8192-tool-use-preview", "llama3-groq-8b-8192-tool-use-preview", "llama3-8b-8192", "llama3-70b-8192"]

--- Models for Cohere
["command-r-plus", "command-r", "command", "command-nightly", "command-light", "command-light-nightly"]
```

Anything else is considered an Ollama model.
Refer to https://ollama.com/library

### CODE REVIEW

File: src/cli/llm/wizard.rs [hunk 1]
```diff
@@ -5,20 +5,19 @@ use genai::resolver::AuthResolver;
 use genai::Client;
 
 use super::Result;
-use crate::cli::llm::model::Model;
 
 #[derive(Setters, Clone)]
 pub struct Wizard<Q, A> {
     client: Client,
-    model: Model,
+    model: String,
```
Comment (tusharmath) [inline src/cli/llm/wizard.rs:12]: We need a type-safe way to configure the LLMs. Please add the `model` back. I think you will need to convert the model into an ENUM.

File: src/cli/generator/config.rs [hunk 2]
```diff
@@ -25,8 +25,16 @@ pub struct Config<Status = UnResolved> {
...
+}
+
+#[derive(Deserialize, Serialize, Debug, Default, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct LLMConfig {
+    pub model: String,
     #[serde(skip_serializing_if = "TemplateString::is_empty")]
-    pub secret: TemplateString,
+    pub api_key: TemplateString,
```
Comment (tusharmath) [inline src/cli/generator/config.rs:37]:
```suggestion
    pub secret: TemplateString,
```

File: src/cli/generator/config.rs [hunk 3]
```diff
@@ -25,8 +25,16 @@ pub struct Config<Status = UnResolved> {
...
     pub preset: Option<PresetConfig>,
     pub schema: Schema,
+    pub llm: LLMConfig,
+}
+
+#[derive(Deserialize, Serialize, Debug, Default, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct LLMConfig {
+    pub model: String,
```
Comment (tusharmath) [inline src/cli/generator/config.rs:35]: Model should be an enum, defined in `llm` module.

File: src/core/mustache/template_string.rs [hunk 4]
```diff
@@ -92,16 +92,16 @@ mod tests {
 
     #[test]
     fn test_parse() {
-        let actual = TemplateString::parse("{{.env.TAILCALL_SECRET}}").unwrap();
-        let expected = Mustache::parse("{{.env.TAILCALL_SECRET}}").unwrap();
+        let actual = TemplateString::parse("{{.env.TAILCALL_LLM_API_KEY}}").unwrap();
+        let expected = Mustache::parse("{{.env.TAILCALL_LLM_API_KEY}}").unwrap();
```
Comment (tusharmath) [inline src/core/mustache/template_string.rs:96]: Are these changes necessary?

Comment (beelchester) [reply src/core/mustache/template_string.rs:96]: No, I used apiKey field name earlier so renamed env also. Reverting this.

File: src/cli/generator/generator.rs [hunk 5]
```diff
@@ -180,13 +180,7 @@ impl Generator {
         let mut config = config_gen.mutation(mutation_type_name).generate(true)?;
 
         if infer_type_names {
```
Comment (tusharmath) [inline src/cli/generator/generator.rs:182]: Also check if LLM model is provided.

File: src/cli/generator/snapshots/tailcall__cli__generator__config__tests__config_codec.snap [hunk 6]
```diff
@@ -16,5 +16,8 @@ expression: actual
     }
   ],
   "output": {},
-  "schema": {}
+  "schema": {},
+  "llm": {
+    "model": ""
+  }
```
Comment (tusharmath) [inline src/cli/generator/snapshots/tailcall__cli__generator__config__tests__config_codec.snap:22]:
```suggestion
  "llm": {}
```

File: src/cli/llm/infer_type_name.rs [hunk 7]
```diff
@@ -3,12 +3,13 @@ use std::collections::HashMap;
...
 use serde::{Deserialize, Serialize};
 
-use super::model::groq;
 use super::{Error, Result, Wizard};
+use crate::cli::generator::config::LLMConfig;
 use crate::core::config::Config;
 
 #[derive(Default)]
 pub struct InferTypeName {
+    model: String,
```
Comment (tusharmath) [inline src/cli/llm/infer_type_name.rs:12]:
```suggestion
    model: Option<String>,
```

File: src/cli/generator/generator.rs [hunk 8]
```diff
@@ -180,13 +180,7 @@ impl Generator {
...
-            let key = if !secret.is_empty() {
-                Some(secret.to_string())
-            } else {
-                None
-            };
-
-            let mut llm_gen = InferTypeName::new(key);
+            let mut llm_gen = InferTypeName::new(llm_config);
```
Comment (tusharmath) [inline src/cli/generator/generator.rs:183]:
```suggestion
            let mut llm_gen = InferTypeName::new(model, key);
```

Pass the model and the key directly from here so that we don't need to verify if the input is valid.

**Review by tusharmath — APPROVED** [review]
